DE-2999: refactor parse logic

setiat opened the pull request:
Changes for `DatetimeInfo`
- `date_str` and `time_str` issue was fixed in previous refactor
- `_date_str` and `_time_str` were helper function used unit tests. I have removed them, test modules will take care of functionality by themselves
- `convert_to_ts_iso8601` was removed as a part of previous ticket
- The major refactor is done for `_parse` and `matcher functions`
- all test cases are passing

setiat commented:
Logic changes for parse:

Right Now the signature of matcher functions looks like:
```python
def matcher_func(self, token:str) -> Dict[str, str]:
    # Logic goes here
```
The return type for each matcher is `Dict[str, str]`, where K,V pair  represent the name of the instance variable to be set  and its value.

The returned `dict` is used in `_parse` to actually [set the value](https://github.com/tetrascience/ts-task-script-utils/blob/DE-2999-refactor-parse-logic/task_script_utils/datetime_parser/datetime_info.py#L73).

Another way of doing this would be that each matcher function will set the value for instance variable and returns `True/False` depending on wether the value was matched or not. For example

```python
def _match_date(self, token:str) -> bool:
    # Logic goes here
    if day and month and year:
        self.day = day
        self.year = year
        self.month = month
        return True
    else:
        return False
```

The is better because this will explicitly set the instance variable. But each matcher will set the the values in its own scope. earlier it was centralised via `_parse`. 

@mlipschultz..Should I change the logic to use 2nd approach?

mlipschultz reviewed (commented):
My concern with the current implementation (where it's centralized in `_parse`) is that it makes tracing through the code with tools (such as IDEs) difficult.  So I like the second approach more since it's easier to trace through the code and setting the values occurs where they're found.

Additional changes requested:
* All of these types will fail a type checker because their types are `str`, but their value is `None`: https://github.com/tetrascience/ts-task-script-utils/blob/master/task_script_utils/datetime_parser/datetime_info.py#L35
    * Perhaps `Optional[str]` would be better

I need to do some more review, but I wanted to save this before I lost it

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -56,68 +56,72 @@ def __str__(self):
+                    _matchers.remove(func)
+                for key, value in result.items():
+                    if self.__dict__.get(key) is None:
+                        self.__dict__[key] = value
```
I have made the changes and every matcher sets the required variables. So these code lines are no longer required in parse.

I am not sure if I can get rid of _parse, even if it is removed...something is still required to iterate the tokens

setiat commented:
> My concern with the current implementation (where it's centralized in `_parse`) is that it makes tracing through the code with tools (such as IDEs) difficult. So I like the second approach more since it's easier to trace through the code and setting the values occurs where they're found.
> 
> Additional changes requested:
> 
> * All of these types will fail a type checker because their types are `str`, but their value is `None`: https://github.com/tetrascience/ts-task-script-utils/blob/master/task_script_utils/datetime_parser/datetime_info.py#L35
>   
>   * Perhaps `Optional[str]` would be better
> 
> I need to do some more review, but I wanted to save this before I lost it

- Added type hints in __init__
- Simplified _parse logic
- Updated matchers to set the instance variable and return True/False depending on wether a match was found or not

mlipschultz commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -61,16 +61,12 @@ def _parse(self, matchers):
-                    if self.__dict__.get(key) is None:
-                        self.__dict__[key] = value
 
     def _parse_long_date_formats(self):
```
The code path that goes through this method looks completely different than the code path through `_parse_short_date_formats`.  This one seems to set tokens to be used by [`_parse_with_formats`](https://github.com/tetrascience/ts-task-script-utils/blob/ebb9bfc14e2d940c20b75298160dd057862feeb9/task_script_utils/datetime_parser/parser.py#L78) whereas `_parse_short_date_formats` seems to set the actual extracted values from the string.  I feel like these two actions are different enough that they should be more separate.  The docstring for `DateTimeInfo` also points to this.  Most of the code in `DateTimeInfo` points to the class doing the parsing and storage of extracted datetime values (i.e. short date).  The long date work doesn't feel like it belongs since it's not doing any of that.  I'd suggest moving the long date code out of this class and move it closer to the pendulum parsing code.

setiat replied:
Create a separate PR for this: #13

mlipschultz reviewed (changes requested):
Additional comments:
* `long datetime_formats`: The [return type hint](https://github.com/tetrascience/ts-task-script-utils/blob/ebb9bfc14e2d940c20b75298160dd057862feeb9/task_script_utils/datetime_parser/datetime_info.py#L507) doesn't match the [type of your return value](https://github.com/tetrascience/ts-task-script-utils/blob/ebb9bfc14e2d940c20b75298160dd057862feeb9/task_script_utils/datetime_parser/datetime_info.py#L515).  `formats` is a generator.
* `_parse_with_formats`: Please add [a more specific exception](https://github.com/tetrascience/ts-task-script-utils/blob/ebb9bfc14e2d940c20b75298160dd057862feeb9/task_script_utils/datetime_parser/parser.py#L103)
* Please add type hints to the return values in https://github.com/tetrascience/ts-task-script-utils/blob/ebb9bfc14e2d940c20b75298160dd057862feeb9/task_script_utils/datetime_parser/ts_datetime.py#L10

mlipschultz commented:
When doing some testing for a proposal (https://tetrascience.atlassian.net/wiki/spaces/DE/pages/3168239653?focusedCommentId=3183673359#comment-3183673359), I discovered a bug:
```python
>>> from task_script_utils.datetime_parser import parse
>>> parse('01/01/0001', ['MM/DD/YYYY']).tsformat()
'1-01-01T00:00:00'
```

The result should be `'0001-01-01T00:00:00'`.  Could you fix this as part of this PR?

setiat commented:
> ```python
> parse('01/01/0001', ['MM/DD/YYYY']).tsformat()
> ```

Fixed it as a part of #13 

I tried fixing this using python datetime  formatting tokens... But [build](https://app.travis-ci.com/github/tetrascience/ts-task-script-utils/builds/247405109) was failing even though tests were passing locally.

So I added custom function _pad_year to fix this

setiat commented on task_script_utils/convert_datetime_to_ts_format.py:
```diff
@@ -1,4 +1,4 @@
-import dateparser, arrow, datetime
+import dateparser, arrow
```
fixed

setiat commented on task_script_utils/datetime_parser/utils.py:
```diff
@@ -116,3 +117,33 @@ def replace_z_with_offset(datetime_str: str) -> str:
+                datetime_str_with_no_abbreviated_tz, format_, tz=None
+            )
+            return parsed, format_
+        except ValueError as e:
```
fixed

setiat commented on task_script_utils/datetime_parser/utils.py:
```diff
@@ -116,3 +117,33 @@ def replace_z_with_offset(datetime_str: str) -> str:
+            return parsed, format_
+        except ValueError as e:
+            continue
+        except re.error as e:
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,414 @@
+
+        if self.abbreviated_tz:
+            if self.abbreviated_tz not in self.config.tz_dict:
+                raise OffsetNotKnownError(
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,414 @@
+    @property
+    def datetime_stamp(self):
+        """Created Datetime string from parsed raw input string.
+        The format is YYYY-MM-DD hh:mm:ss and fractional seconds (upto 6 digit), AM/PM
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,414 @@
+            if self.fractional_seconds:
+                # This property along with dt_format is used to create
+                # a datetime object. Since python datetime support 6 digits for
+                # microseconds, therefore truncating fraction seconds to 6 digits
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,414 @@
-                "abbreviated_tz": self._match_tz_abbreviation,
-            }
+        if not (self.token_month and self.token_day):
+            raise InvalidDateError(f"{self.date_time_raw}")
```
fixed

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+        result = date_info.datetime
+        if result is not None:
+            result = result.isoformat()
+    except DatetimeParserError as e:
```
fixed

setiat commented on __tests__/unit/test_ts_datetime.py:
```diff
@@ -63,3 +73,12 @@ def test_subseconds(input_, iso_format, datetime_iso_format):
+    "input_, format_, expected_ts_format, expected_iso_format", padding_test_cases
+)
+def test_datetime_padding(input_, format_, expected_ts_format, expected_iso_format):
+    ts_datetime = from_pendulum_format(input_, format_, None, None)
```
fixed

setiat commented on __tests__/unit/test_short_datetime_info.py:
```diff
@@ -71,31 +71,38 @@
     try:
-        d = DateTimeInfo(input, year_first)
-        parsed_datetime = d.dtstamp
+        d = ShortDateTimeInfo(input_, year_first)
```
fixed

setiat commented on __tests__/unit/test_short_datetime_info.py:
```diff
@@ -71,31 +71,38 @@
     assert parsed_datetime == expected
+
+
+def _build_date_str_from_datetime_info(dt_info: ShortDateTimeInfo) -> str:
```
fixed

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+        result = date_info.datetime
+        if result is not None:
+            result = result.isoformat()
+    except DatetimeParserError as e:
```
fixed

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+        result = date_info.datetime
+        if result is not None:
+            result = result.isoformat()
+    except DatetimeParserError as e:
```
fixed

setiat commented on __tests__/unit/test_ts_datetime.py:
```diff
@@ -63,3 +73,12 @@ def test_subseconds(input_, iso_format, datetime_iso_format):
+    "input_, format_, expected_ts_format, expected_iso_format", padding_test_cases
+)
+def test_datetime_padding(input_, format_, expected_ts_format, expected_iso_format):
+    ts_datetime = from_pendulum_format(input_, format_, None, None)
```
fixed

setiat commented on task_script_utils/convert_datetime_to_ts_format.py:
```diff
@@ -1,4 +1,4 @@
-import dateparser, arrow, datetime
+import dateparser, arrow
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -132,17 +465,17 @@ def _match_time(self, token: str) -> Optional[dict]:
+            return False
         """
-        hh_mm_ss_pattern = r"\d{1,2}:\d{1,2}:\d{1,2}\.\d+|^\d{1,2}:\d{1,2}:\d{1,2}$|^\d{1,2}:\d{1,2}:\d{1,2}[+-]{1,1}"
+        hh_mm_ss_pattern = r"\d{1,2}:\d{1,2}:\d{1,2}\.\d+|^\d{1,2}:\d{1,2}:\d{1,2}$|^\d{1,2}:\d{1,2}:\d{1,2}[+-]"
```
ignored with inline comment

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+    "Sunday, May 26th 2013 12:12:12.0001 AM Asia/Kolkata": "2013-05-26T00:12:12.0001+05:30",
+    "Sunday, May 26th 2013 12:12:12.0001 AM Asia/Kolkata IST": "2013-05-26T00:12:12.0001+05:30",
+    "Sunday, May 26th 2013 12:12:12 AM IST": "2013-05-26T00:12:12+05:30",
+    "Sunday, May 26th 2013 12:12:12.0001 AM Asia/Kolkata +05:30": "2013-05-26T00:12:12.0001+05:30",
```
ignored with inline comment

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+    "Sunday, May 26 2013 12:12:12.0000 AM Asia/Kolkata": "2013-05-26T00:12:12.0000+05:30",
+    "Sunday, May 26th 2013 12:12:12.5677 AM Asia/Kolkata": "2013-05-26T00:12:12.5677+05:30",
+    "Sunday, May 26th 2013 12:12:12.0001 AM Asia/Kolkata": "2013-05-26T00:12:12.0001+05:30",
+    "Sunday, May 26th 2013 12:12:12.0001 AM Asia/Kolkata IST": "2013-05-26T00:12:12.0001+05:30",
```
ignored with inline comment

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+    "Sunday, May 26 2013 12:12:12.5677 AM Asia/Kolkata": "2013-05-26T00:12:12.5677+05:30",
+    "Sunday, May 26 2013 12:12:12.0000 AM Asia/Kolkata": "2013-05-26T00:12:12.0000+05:30",
+    "Sunday, May 26th 2013 12:12:12.5677 AM Asia/Kolkata": "2013-05-26T00:12:12.5677+05:30",
+    "Sunday, May 26th 2013 12:12:12.0001 AM Asia/Kolkata": "2013-05-26T00:12:12.0001+05:30",
```
ignored with inline comment

setiat replied:
ignored with inline comment

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+    "Sunday, May 26 2013 12:12:12 AM Asia/Kolkata": "2013-05-26T00:12:12+05:30",
+    "Sunday, May 26 2013 12:12:12.5677 AM Asia/Kolkata": "2013-05-26T00:12:12.5677+05:30",
+    "Sunday, May 26 2013 12:12:12.0000 AM Asia/Kolkata": "2013-05-26T00:12:12.0000+05:30",
+    "Sunday, May 26th 2013 12:12:12.5677 AM Asia/Kolkata": "2013-05-26T00:12:12.5677+05:30",
```
ignored with inline comment

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+    "Sunday, May 26th, 2013 12:12:12 AM Asia/Kolkata": "2013-05-26T00:12:12+05:30",
+    "Sunday, May 26 2013 12:12:12 AM Asia/Kolkata": "2013-05-26T00:12:12+05:30",
+    "Sunday, May 26 2013 12:12:12.5677 AM Asia/Kolkata": "2013-05-26T00:12:12.5677+05:30",
+    "Sunday, May 26 2013 12:12:12.0000 AM Asia/Kolkata": "2013-05-26T00:12:12.0000+05:30",
```
ignored with inline comment

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+    "Sunday, May 26th 2013 12:12:12 AM Asia/Kolkata": "2013-05-26T00:12:12+05:30",
+    "Sunday, May 26th, 2013 12:12:12 AM Asia/Kolkata": "2013-05-26T00:12:12+05:30",
+    "Sunday, May 26 2013 12:12:12 AM Asia/Kolkata": "2013-05-26T00:12:12+05:30",
+    "Sunday, May 26 2013 12:12:12.5677 AM Asia/Kolkata": "2013-05-26T00:12:12.5677+05:30",
```
ignored with inline comment

setiat commented on task_script_utils/datetime_parser/parser.py:
```diff
@@ -43,26 +37,17 @@ def parse(
-        parsed_datetime, matched_format = _parse_with_formats(
-            datetime_str, datetime_config=config, formats=formats_list
-        )
+        parsed_datetime, _ = parse_with_formats(datetime_str, config, formats_list)
```
max length changed to 88

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,414 @@
+        ]
+
+        def map_am_pm(time_format):
+            return time_format if time_format.startswith("H") else time_format + " A"
```
max length changed to 88

setiat commented on __tests__/unit/test_short_datetime_info.py:
```diff
@@ -71,31 +71,38 @@
 
 @pytest.mark.parametrize(
-    "input, year_first, day_first, expected", two_digit_date_with_config_test_cases
+    "input_, year_first, day_first, expected", two_digit_date_with_config_test_cases
```
max length changed to 88

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+import pytest
+from task_script_utils.datetime_parser.datetime_info import LongDateTimeInfo
+from task_script_utils.datetime_parser.datetime_config import DatetimeConfig
+from task_script_utils.datetime_parser.parser_exceptions import DatetimeParserError
```
max length changed to 88

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+
+
+long_datetime_test_cases = {
+    "Sunday, May 26th 2013 12:12:12 AM IST Asia/Kolkata": "2013-05-26T00:12:12+05:30",
```
max length changed to 88

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+    "Sunday, May 26th 2013 12:12:12 AM ZST Asia/Kolkata": None,
+    "Sunday, May 26th 2013 12:12:12 AM BST": "2013-05-26T00:12:12+01:00",
+    "Sunday, May 26th 2013 12:12:12.00100 AM BST": "2013-05-26T00:12:12.00100+01:00",
+    "Sunday, May 26th 2013 12:12:12 AM Asia/Kolkata": "2013-05-26T00:12:12+05:30",
```
max length changed to 88

setiat commented on __tests__/unit/test_long_datetime_info.py:
```diff
@@ -0,0 +1,37 @@
+    "Sunday, May 26th 2013 12:12:12 AM BST": "2013-05-26T00:12:12+01:00",
+    "Sunday, May 26th 2013 12:12:12.00100 AM BST": "2013-05-26T00:12:12.00100+01:00",
+    "Sunday, May 26th 2013 12:12:12 AM Asia/Kolkata": "2013-05-26T00:12:12+05:30",
+    "Sunday, May 26th, 2013 12:12:12 AM Asia/Kolkata": "2013-05-26T00:12:12+05:30",
```
max length changed to 88

setiat commented on __tests__/unit/test_datetime_parser_utils.py:
```diff
@@ -0,0 +1,32 @@
+format_list_with_no_tz_dict_test_cases = {
+    "Sunday, May 26th 2013 12:12:12 AM IST Asia/Kolkata": None,
+    "Sunday, May 26th 2013 12:12:12 AM ZST Asia/Kolkata": None,
+    "Sunday, May 26th 2013 12:12:12 AM Asia/Kolkata": "2013-05-26T00:12:12+05:30",
```
max length changed to 88

setiat commented on __tests__/unit/test_datetime_parser_utils.py:
```diff
@@ -0,0 +1,32 @@
+    "input_, expected", format_list_with_no_tz_dict_test_cases.items()
+)
+def test_parse_with_formats_with_no_tz_dict(input_, expected):
+    parsed_datetime, _ = parse_with_formats(input_, DatetimeConfig(), formats_list)
```
max length changed to 88

setiat commented on __tests__/unit/test_ts_datetime.py:
```diff
@@ -63,3 +73,12 @@ def test_subseconds(input_, iso_format, datetime_iso_format):
+
+
+@pytest.mark.parametrize(
+    "input_, format_, expected_ts_format, expected_iso_format", padding_test_cases
```
max length changed to 88

setiat commented on __tests__/unit/test_ts_datetime.py:
```diff
@@ -63,3 +73,12 @@ def test_subseconds(input_, iso_format, datetime_iso_format):
+@pytest.mark.parametrize(
+    "input_, format_, expected_ts_format, expected_iso_format", padding_test_cases
+)
+def test_datetime_padding(input_, format_, expected_ts_format, expected_iso_format):
```
max length changed to 88

setiat commented on __tests__/unit/test_short_datetime_info.py:
```diff
@@ -71,31 +71,38 @@
-@pytest.mark.parametrize("input, expected", regex_test_cases)
-def test_regex_parsing(input, expected):
+@pytest.mark.parametrize("input_, expected", regex_test_cases)
+def test_regex_parsing(input_, expected):
```
ignored

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -132,17 +421,17 @@ def _match_time(self, token: str) -> Optional[dict]:
+            return False
         """
-        hh_mm_ss_pattern = r"\d{1,2}:\d{1,2}:\d{1,2}\.\d+|^\d{1,2}:\d{1,2}:\d{1,2}$|^\d{1,2}:\d{1,2}:\d{1,2}[+-]{1,1}"
+        hh_mm_ss_pattern = r"\d{1,2}:\d{1,2}:\d{1,2}\.\d+|^\d{1,2}:\d{1,2}:\d{1,2}$|^\d{1,2}:\d{1,2}:\d{1,2}[+-]"
```
max length changed to 88

setiat commented on task_script_utils/datetime_parser/utils.py:
```diff
@@ -116,3 +117,33 @@ def replace_z_with_offset(datetime_str: str) -> str:
     return re.sub(r"(?<=\d|\s)Z(?=\s|$)", "+00:00", datetime_str)
+
+
+def parse_with_formats(
```
added doc strings

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,414 @@
+        # Input =  2018-13-09T11:12:23.000-05:30
+        # output = 2018-13-09 11:12:23.000-05:30
+        char_list = list(string)
+        for i, current_char in enumerate(char_list[1:-1]):
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,414 @@
+        processed_dt = self._remove_T_between_two_digits(raw_dt)
+        return processed_dt.split()
+
+    def _remove_T_between_two_digits(self, string: str):
```
fixed

setiat commented on task_script_utils/datetime_parser/utils.py:
```diff
@@ -116,3 +117,44 @@ def replace_z_with_offset(datetime_str: str) -> str:
+        formats (Sequence[str], optional): list of formats. Defaults to ().
+
+    Returns:
+        Tuple[Optional[TSDatetime], Optional[str]]: return TSDatetime, matched_format
```
max Len changed to 88

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,426 @@
 from .tz_list import _all_abbreviated_tz_list
+from .ts_datetime import TSDatetime
+from .utils import parse_with_formats
+from .parser_exceptions import (
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,426 @@
 from .tz_list import _all_abbreviated_tz_list
+from .ts_datetime import TSDatetime
+from .utils import parse_with_formats
+from .parser_exceptions import (
```
fixed

mlipschultz commented on __tests__/unit/test_datetime_parser_utils.py:
```diff
@@ -0,0 +1,32 @@
+import pytest
```
@abclay, what's the proper solution here?  Tarik disabled the error, but I don't know if it's the best approach.  I'm also not sure why reviewdog gave this error.

abclay commented:
I just want to point out that fixing/resolving all the pylint/flake8 errors is not necessary to merge the PR.  They are there more for information than for blocking.  I know it's overly verbose right now, so I apologize for that.  If you have feedback, I would recommend adding it to this doc or its comments:  https://tetrascience.atlassian.net/wiki/spaces/DE/pages/3164700810/Linting+Ignore+List+Proposals

mlipschultz commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -39,6 +58,9 @@ def __str__(self):
         return json.dumps(self.__dict__, indent=2)
 
     def _parse(self, matchers):
```
I haven't had time to dig into whether this pattern still makes sense for the new parsing strategy (with two classes).  I think that's the only thing left for me to look at.  Unfortunately, I won't be able to review until Tuesday so maybe someone else can do that part of the review.

mlipschultz replied:
I had a chance to look through this.  I don't like that the ShortDateTimeInfo matchers extract actual values while the LongDateTimeInfo matchers identify what token to use.  However, the outcome from the caller's perspective is the same, so I think I'm ok with it.

mlipschultz reviewed (commented):
I mostly have small requests.  I wasn't able to finish my review before ending for the day, but there's just one last thing I was hoping to look at.  Maybe someone else can review that part?

abclay commented on .pylintrc:
```diff
@@ -0,0 +1,24 @@
+[MASTER]
```
Yes, this looks fine to me.

setiat commented on task_script_utils/convert_datetime_to_ts_format.py:
```diff
@@ -30,17 +38,17 @@ def convert_datetime_to_ts_format(
         try:
             timezone_to_use = tz.gettz(timezone)
-        except:
+        except Exception:
```
I dont think `gettz` will ever raise except.. I tried passing gibberish to it and it returned None.

Can I create a separate ticket on fixing the old code? At this moment can't say how  changing code will impact behaviour in prod.

mlipschultz replied:
I forgot to mention that I'm not opposed to having this be a separate ticket, but I don't think I'd want to release until we decide about what to do when `None` is returned on bad input.

mlipschultz replied:
When `None` is returned, there's [code later](https://github.com/tetrascience/ts-task-script-utils/blob/d841e2cdac3412743f27036ed9c4b4bffdaa5d9e/task_script_utils/convert_datetime_to_ts_format.py#L66) to handle a false-y value -- it'll default to GMT.  However, is this the behavior we want on unrecognized timezones?  This feels a little like hiding an error in the input, which could be a data integrity issue ([something we're trying to address right now elsewhere in our codebase](https://tetrascience.slack.com/archives/GMTG3FDNH/p1647259200879819)).  @robjhornby-ts, @wkopen, what are your thoughts here?

To summarize: If `tz.gettz` receives a string that can't be recognized as a timezone, it returns `None`.  Later in the code, if the `timezone_to_use` is false-y, we convert it to GMT without logging a warning/error (and without stopping parsing and raising an exception).  Is this behavior ok?  My guess is no, but I'd like confirmation

mlipschultz replied:
Also, @setiat, I tihnk we talked about converting this `Exception` to `AttributeError` because that's the only exception we know should occur (when a non-string is passed in).

setiat replied:
@mlipschultz and @robjhornby-ts 

`timezone` passed to `convert_datetime_to_ts_format` could be a `str`(valid or invalid) or it could be a `tz.offset`.  In case it is `tz.offset`, a `TypeError` is raised and code still works

If I pass an `int` as `timezone` and catch only `AttributeError`, code will still fail.

So, I have made following changes
- Updated typehints for `str` to `Union[tz.offset, str]`
- Handle only `TypeError`, every other exception will be raised
- added a test case

setiat commented on task_script_utils/convert_datetime_to_ts_format.py:
```diff
@@ -30,17 +38,19 @@ def convert_datetime_to_ts_format(
             timezone_to_use = tz.gettz(timezone)
-        except:
+        except Exception:
             print(
```
Are print statements logged by platform?

robjhornby-ts replied:
I think this file has only changed because of pylint problems, is that right? I agree it would be nice to remove `print` from this function and use something more flexible, but I don't know what that would look like yet so maybe we could leave it for a future ticket.

It would be good to have a general way of logging / returning status messages which can easily be displayed as logs in the platform OR can easily be ignored by the caller, depending on what's wanted for that specific task script.

Would you rather replace it or remove it as part of this PR @mlipschultz ?

setiat replied:
added `warnings.warn` instead of `print`

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -746,46 +799,34 @@ def _process_day_and_month(self, tokens: List[str]) -> tuple:
 
-    def _pad_and_validate_offset_value(self, offset):
+    @staticmethod
+    def _pad_and_validate_offset_value(offset):
```
renamed the functions to :`_pad_and_validate_time_offset_value `

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -704,14 +756,15 @@ def _process_two_digit_date_pattern(self, date_parts):
 
-    def _process_day_and_month(self, tokens: List[str]) -> tuple:
+    @staticmethod
+    def _process_day_and_month(tokens: List[str]) -> tuple:
```
Fixed the name and function signature. It know access `first_token` and `second_token` instead of a list

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -704,14 +756,15 @@ def _process_two_digit_date_pattern(self, date_parts):
-            dt_date = dt(day=int(day), month=int(month), year=int(year))
-        except Exception as e:
-            msg = f"{str(e)}, date={'-'.join(date_parts)}, {self.config}"
+            _ = dt(day=int(day), month=int(month), year=int(year))
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -641,16 +692,17 @@ def _process_year_first_or_last_matches(self, date_parts, year_first):
-            dt_date = dt(day=int(day), month=int(month), year=int(year))
-        except Exception as e:
-            msg = f"{str(e)}, date={date_parts}, config={self.config}"
+            _ = dt(day=int(day), month=int(month), year=int(year))
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,464 @@
+from datetime import datetime as dt
 from itertools import product
-from typing import Dict, List, Optional, Tuple, Union
+from typing import List, Optional, Tuple
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -608,88 +756,72 @@ def _process_two_digit_date_pattern(self, date_parts):
 
         if first_token_is_month and second_token_is_month:
-            raise AmbiguousDateError(f"Can't decide day and month between: {tokens}")
+            raise AmbiguousDateError(f"Can't decide day and month between: {first_token}, {second_token}")
```
fixed using black

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -608,88 +756,72 @@ def _process_two_digit_date_pattern(self, date_parts):
 
         if not first_token_is_month and not second_token_is_month:
-            raise AmbiguousDateError(f"Can't decide day and month between: {tokens}")
+            raise AmbiguousDateError(f"Can't decide day and month between: {first_token}, {second_token}")
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -608,88 +756,72 @@ def _process_two_digit_date_pattern(self, date_parts):
 
         day, month = (
-            (tokens[0], tokens[1]) if second_token_is_month else (tokens[1], tokens[0])
+            (first_token, second_token) if second_token_is_month else (second_token, first_token)
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -608,88 +756,72 @@ def _process_two_digit_date_pattern(self, date_parts):
 
         if first_token_is_month and second_token_is_month:
-            raise AmbiguousDateError(f"Can't decide day and month between: {tokens}")
+            raise AmbiguousDateError(f"Can't decide day and month between: {first_token}, {second_token}")
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -608,88 +756,72 @@ def _process_two_digit_date_pattern(self, date_parts):
 
         if not first_token_is_month and not second_token_is_month:
-            raise AmbiguousDateError(f"Can't decide day and month between: {tokens}")
+            raise AmbiguousDateError(f"Can't decide day and month between: {first_token}, {second_token}")
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -608,88 +756,72 @@ def _process_two_digit_date_pattern(self, date_parts):
 
         day, month = (
-            (tokens[0], tokens[1]) if second_token_is_month else (tokens[1], tokens[0])
+            (first_token, second_token) if second_token_is_month else (second_token, first_token)
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -1,125 +1,464 @@
+from datetime import datetime as dt
 from itertools import product
-from typing import Dict, List, Optional, Tuple, Union
+from typing import List, Optional, Tuple
```
fixed

mlipschultz reviewed (commented):
I have just two small docstring typos to fix.  Once the `print` statement is switched, I can approve

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -756,51 +756,52 @@ def _process_two_digit_date_pattern(self, date_parts):
-            tokens (List): a list containing two
-            numeric tokens
+    def _disambiguate_day_and_month(first_token: str, second_token: str) -> tuple:
+        """Takes two tokens as input and try to
```
fixed

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -756,51 +756,52 @@ def _process_two_digit_date_pattern(self, date_parts):
 
         Raises:
-            AmbiguousDateError: When we fail to decide between day and month
+            AmbiguousDateError: When it fail to decide between day and month
```
fixed

mlipschultz reviewed (commented):
Everything looks good.  There's still the question on what to do with non-timezone strings, but I'd like to get the perspective of @robjhornby-ts or @wkopen since they have more context.

setiat commented on task_script_utils/datetime_parser/datetime_info.py:
```diff
@@ -545,8 +689,8 @@ def _process_year_first_or_last_matches(self, date_parts, year_first):
-            dt_date = dt(day=int(day), month=int(month), year=int(year))
-        except Exception as e:
+            _ = dt(day=int(day), month=int(month), year=int(year))
+        except ValueError as e:
```
fixed.

setiat commented on task_script_utils/convert_datetime_to_ts_format.py:
```diff
@@ -30,19 +39,23 @@ def convert_datetime_to_ts_format(
+                "The provided timezone can't be parsed by dateutil tz, going to use the plain value"  # noqa: E501
             )
             timezone_to_use = timezone
+        except AttributeError:
```
fixed

setiat commented on __tests__/unit/test_convert_datetime_to_ts_format.py:
```diff
@@ -1,5 +1,7 @@
+from cmath import exp
```
fixed.

setiat commented on __tests__/unit/test_convert_datetime_to_ts_format.py:
```diff
@@ -33,9 +35,15 @@ def test_input_timezone():
     ]
     for raw, timezone, expect in raw_timezone_expect:
-        assert convert_datetime_to_ts_format(raw, timezone=timezone) == expect
+        if type(expect) is str:
```
fixed

mlipschultz commented on .travis.yml:
```diff
@@ -2,7 +2,7 @@ language: python
   - "3.7"
 install:
-  - pip install pipenv
+  - pip install pipenv==2022.1.8
```
What's the reason for this specific version?

setiat replied:
Travis was struggling to run `Pipenv install --dev`. This is the fix for that

mlipschultz reviewed (commented):
Thanks for switching to `TypeError`!

wkopen commented:
> Everything looks good. There's still the question on what to do with non-timezone strings, but I'd like to get the perspective of @robjhornby-ts or @wkopen since they have more context.

I don't think we should assume it to be "GMT", if timezone info is not available, we don't put timezone in the parsed string. And we def don't want to by-default using "GMT" without logging anything.

robjhornby-ts commented:
> Everything looks good. There's still the question on what to do with non-timezone strings, but I'd like to get the perspective of @robjhornby-ts or @wkopen since they have more context.

> When None is returned, there's [code later](https://github.com/tetrascience/ts-task-script-utils/blob/d841e2cdac3412743f27036ed9c4b4bffdaa5d9e/task_script_utils/convert_datetime_to_ts_format.py#L66) to handle a false-y value -- it'll default to GMT. However, is this the behavior we want on unrecognized timezones? This feels a little like hiding an error in the input, which could be a data integrity issue ([something we're trying to address right now elsewhere in our codebase](https://tetrascience.slack.com/archives/GMTG3FDNH/p1647259200879819)). @robjhornby-ts, @wkopen, what are your thoughts here?
> 
> To summarize: If tz.gettz receives a string that can't be recognized as a timezone, it returns None. Later in the code, if the timezone_to_use is false-y, we convert it to GMT without logging a warning/error (and without stopping parsing and raising an exception). Is this behavior ok? My guess is no, but I'd like confirmation

The default to GMT logic is a bit misleading, that branch of logic will lead to a datetime output with no timezone (`utc_indicator = ""`). It's a workaround because `arrow` and `dateparser` didn't have the functionality we needed.

The `convert_datetime_to_ts_format` function should be deprecated now that the new TS datetime parse function is ready to go. Or we should change `convert_datetime_to_ts_format` to use the new TS datetime parse function instead. Right now `convert_datetime_to_ts_format` has several problems which is why we wanted the new TS datetime library.

mlipschultz commented:
> The default to GMT logic is a bit misleading, that branch of logic will lead to a datetime output with no timezone (`utc_indicator = ""`). It's a workaround because `arrow` and `dateparser` didn't have the functionality we needed.

Thanks for the context.  If we keep the code, I think we should leave a comment so others understand in the future.  That said, ...

> The `convert_datetime_to_ts_format` function should be deprecated now that the new TS datetime parse function is ready to go. Or we should change `convert_datetime_to_ts_format` to use the new TS datetime parse function instead. Right now `convert_datetime_to_ts_format` has several problems which is why we wanted the new TS datetime library.

I think we should change the implementation to use the new TS datetime parser function -- no point in duplicating code.

setiat commented:
> > The default to GMT logic is a bit misleading, that branch of logic will lead to a datetime output with no timezone (`utc_indicator = ""`). It's a workaround because `arrow` and `dateparser` didn't have the functionality we needed.
> 
> Thanks for the context. If we keep the code, I think we should leave a comment so others understand in the future. That said, ...
> 
> > The `convert_datetime_to_ts_format` function should be deprecated now that the new TS datetime parse function is ready to go. Or we should change `convert_datetime_to_ts_format` to use the new TS datetime parse function instead. Right now `convert_datetime_to_ts_format` has several problems which is why we wanted the new TS datetime library.
> 
> I think we should change the implementation to use the new TS datetime parser function -- no point in duplicating code.

I have created a separate ticket for this effort --> https://tetrascience.atlassian.net/browse/DE-3634

mlipschultz reviewed (approved):
Thanks!
